File: leetcode/400/lc474findMaxForm.go
func findMaxForm(strs []string, m int, n int) int {
	dp := make([][]int, m+1)
	for i := range dp {
		dp[i] = make([]int, n+1)
	}

	for _, cur := range strs {
		cnt0, cnt1 := 0, 0
		for _, ch := range cur {
			if ch == '0' {
				cnt0++
			} else {
				cnt1++
			}
		}

		for i := m; i >= cnt0; i-- {
			for j := n; j >= cnt1; j-- {
				dp[i][j] = max(dp[i][j], dp[i-cnt0][j-cnt1]+1)
			}
		}
	}

	return dp[m][n]
}

// max 返回两个整数中的最大值
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: leetcode/400/LC474findMaxForm.java
public class LC474findMaxForm {

    public int findMaxForm(String[] strs, int m, int n) {
        int[][] funcPre = new int[m + 1][n + 1];
        int[][] funcCur = new int[m + 1][n + 1];

        for (String str : strs) {
            funcCur = combine(funcPre, str);
            funcPre = funcCur;
        }

        return funcPre[m][n];
    }

    private int[][] combine(int[][] funcPre, String str) {
        int[][] funcCur = new int[funcPre.length][funcPre[0].length];

        int cnt0 = (int) str.chars().filter(ch -> ch == '0').count();
        int cnt1 = (int) str.chars().filter(ch -> ch == '1').count();

        for (int i = 0; i < funcCur.length; i++) {
            for (int j = 0; j < funcCur[0].length; j++) {
                if(i >= cnt0 && j>= cnt1){
                    funcCur[i][j] = Math.max(funcPre[i][j], funcPre[i - cnt0][j - cnt1] + 1);
                }else {
                    funcCur[i][j] = funcPre[i][j];
                }
            }
        }
        return funcCur;
    }
}
